### ISSUE TITLE
Linear backward lost grad

### ISSUE THREAD

**Issue by ailzhang** [pytorch/xla#908]
```
import torch
import torch_xla
import torch.nn as nn
import pdb

device = 'xla:0'
# device='cpu'
m = nn.Linear(10, 8).to(device)
input = torch.randn(4, 10, requires_grad=True).to(device)
output = m(input)
output.sum().backward()
print(input.grad)
```
Expect `input.grad` to be a tensor but it's now None.

**Comment by ailzhang** [comment]
Hmmm this sounds like a serious bug to me, I'm going to spend some time on this first, to identify the underlying problem. (Maybe it's related to the `flatten()` issue @zcain117 and @dlibenzi are investigating, let me double check).

**Comment by dlibenzi** [comment]
I have a strong feeling it is @ailzhang :)

**Comment by ailzhang** [comment]
```
import torch_xla
import torch.nn as nn
import pdb

device = 'xla:0'
device='cpu'
# input = torch.randn(4, 10, requires_grad=True, device=device)
input = torch.randn(4, 10, requires_grad=True).to(device)
print(input.is_leaf)
```

**Comment by ailzhang** [comment]
@dlibenzi sorry about the false alarm, this is indeed a wrong way to test. 
So the right way to test is 
```
import torch_xla
import torch.nn as nn
import pdb

device = 'xla:0'
# device='cpu'
m = nn.Linear(10, 8).to(device)
input = torch.randn(4, 10, requires_grad=True, device=device)
output = m(input)
output.sum().backward()
print(input.grad)
```
And xla does pass the test. 
The key difference is I used 
`input = torch.randn(4, 10, requires_grad=True).to(device)` which made `input` a non-leaf variable in terms of autograd, so it should have grad. 
I'm closing this issue since we do have the correct behavior.

**Comment by zcain117** [comment]
I noticed that it seems to matter whether the original variable was initialized on device or copied there.

If we initialize t1 like @ailzhang did above, t1.grad is None:
```
    xla_device = xm.xla_device()
    import pdb; pdb.set_trace()
    t1 = torch.tensor([[2., 2.], [2., 2.]], requires_grad=True).to(xla_device)
    m = nn.Linear(4, 10).to(xla_device)
    viewed = t1.view(-1, 4)
    viewed_linear = m(viewed)
    viewed_sum = viewed_linear.sum()
    grad = torch.rand(1)
    viewed_sum.backward(grad.to(xla_device))
```

If the requires_grad arg is attached to the CPU variable instead of the return value of copy() like below, then x.grad is populated and t1.grad is None:
```
    xla_device = xm.xla_device()
    import pdb; pdb.set_trace()
    x = torch.tensor([[2., 2.], [2., 2.]], requires_grad=True)
    t1 = x.to(xla_device)
    m = nn.Linear(4, 10).to(xla_device)
    viewed = t1.view(-1, 4)
    viewed_linear = m(viewed)
    viewed_sum = viewed_linear.sum()
    grad = torch.rand(1)
    viewed_sum.backward(grad.to(xla_device))
```

If we initialize using 'device' keyword arg, then t1 has non-None grad:
```
    xla_device = xm.xla_device()
    import pdb; pdb.set_trace()
    t1 = torch.tensor([[2., 2.], [2., 2.]], requires_grad=True, device=xla_device)
    m = nn.Linear(4, 10).to(xla_device)
    viewed = t1.view(-1, 4)
    viewed_linear = m(viewed)
    viewed_sum = viewed_linear.sum()
    grad = torch.rand(1)
    viewed_sum.backward(grad.to(xla_device))
```

**Comment by ailzhang** [comment]
@zcain117 Yup and these are expected behavior. 
When you initialize on CPU and then copy to xla, the cpu tensor is a leaf node (in terms of autograd), so all the grad get populated back to CPU tensor. 
When you initialize the tensor with `(..., device='xla')` keyword, the xla tensor is a leaf node(in terms of autograd), so all the grad get populated back to XLA tensor. 

In Pytorch we only populate gradient back to the bottom layer of the graph, which means the first tensor that requires grad (is_leaf = True). :D
